apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,323 @@
+* Decompose operators using other operators for which we already have a
+batching rule. See for example the [batching rule for torch.vdot](https://github.com/pytorch/pytorch/blob/b30ee35a6f141d3247a24fd09f96ea50a7e2b3c7/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp#L35-L37)
+
+## Composite Compliance
+
+As mentioned earlier, we obtain batching rules effortlessly for composite operators
+but this holds true only under certain constraints. These constraints include
+refraining from accessing the tensor's data pointer and avoiding the use of
+`out=` variants of the operators.
+For the full list of constraints, see [this documentation](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md#composite-compliance).
+When all these hold for an operator, we say that it is composite compliant.
+
+Unfortunately, operators that claim to be composite may occasionally deviate
+from these constraints. While such deviations may not pose issues when utilizing
+plain eager PyTorch, they can lead to complications when using `torch.func`
+transformations.
+
+### Testing for Composite Compliance
+
+We have tests to ensure that operators tagged as composite are indeed composite
+compliant. We test this by creating a new subclass `CompositeCompliantTensor`
+that utilizes the `__torch_dispatch__` mechanism. This mechanism is invoked for
+all operators in the testing, enabling us to detect any non-compliant behavior
+exhibited by an operator.
+
+Our testing approach involves running tests on the actual [operator](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1446),
+as well as their [backward formula](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1459)
+and [forward AD formula](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1476).
+Testing both the backward and forward formulas is crucial because we may encounter
+scenarios involving `vmap(vjp(fn))` or `vmap(jvp(fn))`.
+
+## Support for `chunk_size` in `vmap` and `jacrev`
+
+The computation of the Jacobian can be memory-intensive, and users have raised
+concerns about related issues, (e.g., [this one](https://github.com/pytorch/functorch/issues/680)).
+In response to these concerns, we have introduced a feature that allows for the
+calculation of `jacrev` and `vmap` in smaller, user-defined chunks, determined
+by the `chunk_size` argument. This adjustment serves to reduce the peak memory
+usage during the computation process. With this argument, users can specify the
+number of rows of the Jacobian to be computed at once. This enhancement was
+incorporated into the [jacrev PR](https://github.com/pytorch/pytorch/pull/89376)
+and the [vmap PR](https://github.com/pytorch/pytorch/pull/91157).
+
+## Support for `linearize` transform
+
+The "jvp" transform is designed to calculate both `f(x)` and the Jacobian-vector
```

Comment:
Is jvp supposed to be quoted here instead of in code formatting?

Reply:
yep, this should be code formatting. Good catch.